test/src/org/apache/jmeter/gui/action/HtmlReportGeneratorSpec.groovy

```diff
@@ -0,0 +1,128 @@
+            ]
+            JMeterUtils.getJMeterBinDir()+"/testfiles/HTMLReportTestFileMissing.csv" | JMeterUtils.getJMeterBinDir()+"/user.properties"          | JMeterUtils.getJMeterBinDir()+"/testfiles/testReport" | [
+                JMeterUtils.getResString("generate_report_ui.csv_file") + MessageFormat.format(JMeterUtils.getResString(HtmlReportGenerator.NO_FILE), csvPath)
+            ]
+            ""                                                                | ""                                                        | ""                                                    | [
+                JMeterUtils.getResString("generate_report_ui.csv_file") + MessageFormat.format(JMeterUtils.getResString(HtmlReportGenerator.NO_FILE), csvPath),
+                JMeterUtils.getResString("generate_report_ui.user_properties_file") + MessageFormat.format(JMeterUtils.getResString(HtmlReportGenerator.NO_FILE), userPropertiesPath),
+                JMeterUtils.getResString("generate_report_ui.output_directory") + MessageFormat.format(JMeterUtils.getResString(HtmlReportGenerator.CANNOT_CREATE_DIRECTORY), outputDirectoryPath)
+            ]
+            JMeterUtils.getJMeterBinDir()+"/testfiles/HTMLReportTestFile.csv" | JMeterUtils.getJMeterBinDir()+"/user.properties" | JMeterUtils.getJMeterBinDir()+"/testfiles/testReport/oneLevel/twolevel"            | [
+                JMeterUtils.getResString("generate_report_ui.output_directory") + MessageFormat.format(JMeterUtils.getResString(HtmlReportGenerator.CANNOT_CREATE_DIRECTORY), outputDirectoryPath)
+            ]
+    }
+    
+    def "check that report generation succeeds and statistic are generated"(){
+        setup:
+            File testDirectory = new File(JMeterUtils.getJMeterBinDir(), "/testfiles/testReport")
+            if(testDirectory.exists()) {
+                if (testDirectory.list().length>0) {
+                    FileUtils.cleanDirectory(testDirectory)
+                }
+            } else {
+                testDirectory.mkdir()
+            }
+            ObjectMapper mapper = new ObjectMapper()
+            File expected = new File(JMeterUtils.getJMeterBinDir(),"/testfiles/HTMLReportExpect.json")
+            JsonNode expectedRoot = null;
+            expected.withReader { jsonFileReader ->
+                expectedRoot = mapper.readTree(jsonFileReader)
+            }
+        when:
+            HtmlReportGenerator htmlReportGenerator = new HtmlReportGenerator(
+                    JMeterUtils.getJMeterBinDir() + "/testfiles/HTMLReportTestFile.csv",
+                    JMeterUtils.getJMeterBinDir() + "/user.properties", 
+                    JMeterUtils.getJMeterBinDir() + "/testfiles/testReport")
+            List<String> resultList = htmlReportGenerator.run()
+            File statistics = new File(JMeterUtils.getJMeterBinDir(), "/testfiles/testReport/statistics.json")
```

Comment:
What's in this file? It is not a part of the PR.

Reply:
The expected json results

Reply:
It is truly puzzling, and it is not clear when you just read the test code.

Frankly speaking, I don't like the idea to store results in `bin/` directory.